core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -167,20 +170,22 @@ public synchronized void connect() throws AlluxioStatusException {
     Preconditions.checkState(!mClosed, "Client is closed, will not try to connect.");
 
     RetryPolicy retryPolicy =
-        new ExponentialBackoffRetry(BASE_SLEEP_MS, MAX_SLEEP_MS, RPC_MAX_NUM_RETRY);
+        TimeBoundedExponentialRetry.builder().withMaxDuration(MAX_RETRY_DURATION)
+            .withInitialSleep(BASE_SLEEP_MS).withMaxSleep(MAX_SLEEP_MS).build();
     while (true) {
       if (mClosed) {
         throw new FailedPreconditionException("Failed to connect: client has been closed");
       }
-      // Re-query the address in each loop iteration in case it has changed (e.g. master failover).
-      mAddress = getAddress();
-      LOG.info("Alluxio client (version {}) is trying to connect with {} @ {}",
-          RuntimeConstants.VERSION, getServiceName(), mAddress);
-
-      TProtocol binaryProtocol =
-          new TBinaryProtocol(mTransportProvider.getClientTransport(mParentSubject, mAddress));
-      mProtocol = new TMultiplexedProtocol(binaryProtocol, getServiceName());
       try {
+        // Re-query the address in each loop iteration in case it has changed (e.g. master
+        // failover). This must happen inside the try since querying the master address can fail.
+        mAddress = getAddress();
+        LOG.info("Alluxio client (version {}) is trying to connect with {} @ {}",
+            RuntimeConstants.VERSION, getServiceName(), mAddress);
+
+        TProtocol binaryProtocol =
+            new TBinaryProtocol(mTransportProvider.getClientTransport(mParentSubject, mAddress));
+        mProtocol = new TMultiplexedProtocol(binaryProtocol, getServiceName());
         mProtocol.getTransport().open();
         LOG.info("Client registered with {} @ {}", getServiceName(), mAddress);
         mConnected = true;
```

Comment:
Do we want to have separate try-catch blocks, since after here we no longer need to reconnect? Also in the catches, do we need to clean up the existing connection if it exists?